pkg/redpanda/client.go

```diff
@@ -0,0 +1,210 @@
+		if len(tokens) != 3 {
+			continue
+		}
+
+		if !slices.Contains(supportedSASLMechanisms, tokens[2]) {
+			continue
+		}
+
+		user, password, mechanism = tokens[0], tokens[1], tokens[2]
+		return
+	}
+
+	return
+}
+
+func saslOpt(user, password, mechanism string) kgo.Opt {
+	var m sasl.Mechanism
+	switch mechanism {
+	case "PLAIN":
+		m = plain.Auth{User: user, Pass: password}.AsMechanism()
+	case "SCRAM-SHA-256", "SCRAM-SHA-512":
+		scram := scram.Auth{User: user, Pass: password}
+
+		switch mechanism {
+		case "SCRAM-SHA-256":
+			m = scram.AsSha256Mechanism()
+		case "SCRAM-SHA-512":
+			m = scram.AsSha512Mechanism()
+		}
+	default:
+		panic(fmt.Sprintf("unhandled SASL mechanism: %s", mechanism))
+	}
+
+	return kgo.SASL(m)
+}
+
+// DialContextFunc is a function that acts as a dialer for the underlying Kafka client.
+type DialContextFunc func(ctx context.Context, network, host string) (net.Conn, error)
+
+func wrapTLSDialer(dialer DialContextFunc, config *tls.Config) DialContextFunc {
+	return func(ctx context.Context, network, host string) (net.Conn, error) {
+		conn, err := dialer(ctx, network, host)
+		if err != nil {
+			return nil, err
+		}
+		return tls.Client(conn, config), nil
+	}
+}
+
+// KafkaClient creates a client to talk to a Redpanda cluster based on its helm
+// configuration over its internal listeners.
+func KafkaClient(config kube.Config, release helmette.Release, partial redpanda.PartialValues, dialer DialContextFunc, opts ...kgo.Opt) (*kgo.Client, error) {
+	dot, err := redpanda.Dot(release, partial)
+	if err != nil {
+		return nil, err
+	}
+	dot.KubeConfig = config
+
+	values := helmette.Unwrap[redpanda.Values](dot.Values)
+
+	name := redpanda.Fullname(dot)
+	namespace := dot.Release.Namespace
+	serviceName := redpanda.ServiceName(dot)
+	saslUsers := redpanda.SecretSASLUsers(dot)
+	clientCertName := fmt.Sprintf("%s-client", name)
+	kafkaRootCertName := fmt.Sprintf("%s-%s-root-certificate", name, values.Listeners.Kafka.TLS.Cert)
+	serverName := fmt.Sprintf("%s.%s.svc", serviceName, namespace)
+
+	brokers := []string{}
+	for i := int32(0); i < values.Statefulset.Replicas; i++ {
+		brokers = append(brokers, fmt.Sprintf("%s-%d.%s.%s.svc:%d", name, i, serviceName, namespace, values.Listeners.Kafka.Port))
+	}
+
+	opts = append(opts, kgo.SeedBrokers(brokers...))
+
+	serverTLSError := func(err error) error {
+		return fmt.Errorf("error fetching server root CA %s/%s: %w", namespace, kafkaRootCertName, err)
+	}
+	clientTLSError := func(err error) error {
+		return fmt.Errorf("error fetching client certificate default/%s: %w", clientCertName, err)
+	}
+	saslError := func(err error) error {
+		return fmt.Errorf("error fetching SASL authentication for %s/%s: %w", saslUsers.Namespace, saslUsers.Name, err)
+	}
+
+	if redpanda.TLSEnabled(dot) {
+		tlsConfig := &tls.Config{MinVersion: tls.VersionTLS12, ServerName: serverName}
+
+		serverCert, found := helmette.Lookup[corev1.Secret](dot, namespace, kafkaRootCertName)
```

Comment:
I don't think we want to use this function outside of helm packages if we can avoid it. The lack of an explicit client specification, panics, and context support make it a poor choice.  We should probably add some redtape around it 😓 

The `kube` package's `Get` function has the same ergonomics albeit you'll need a `kube.Ctl` instance

Reply:
Just to keep the semantics fairly similar, I introduced a `SafeLookup` in the package that doesn't panic, the `Lookup` then follows more or less the behavior of a `Must*` function, it just calls `SafeLookup` and panics on any returned errors.